resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/tpr/TPRCrudCallsImpl.scala

```diff
@@ -0,0 +1,320 @@
+    val url = generateHttpUrl(requestSegments, Seq(("watch", "true")))
+
+    val request = completeRequest(new Request.Builder()
+      .get()
+      .url(url))
+
+    logDebug(s"Watch Request: $request")
+    val resp = watchClient.newCall(request).execute()
+    completeRequestWithExceptionIfNotSuccessful(
+      "start watch on",
+      resp,
+      Option(Seq(resp.code().toString, resp.message())))
+
+    logInfo(s"Starting watch on jobResource")
+    watchJobObjectUtil(resp)
+  }
+
+/**
+ * This method has a blocking call - wait on SSE - inside it.
+ * However it is sent off in a new thread
+ */
+  private def watchJobObjectUtil(response: Response): Future[WatchObject] = {
+    @volatile var wo: WatchObject = null
+    watchSource = response.body().source()
+    executeBlocking {
+      breakable {
+        // This will block until there are bytes to read or the source is exhausted.
+        while (!watchSource.exhausted()) {
+          watchSource.read(buffer, 8192) match {
+            case -1 =>
+              cleanUpListener(watchSource, buffer)
+              throw new SparkException("Source is exhausted and object state is unknown")
+            case _ =>
+              wo = read[WatchObject](buffer.readUtf8())
+              wo match {
+                case WatchObject("DELETED", w) =>
+                  logInfo(s"${w.metadata.name} has been deleted")
+                  cleanUpListener(watchSource, buffer)
+                case WatchObject(e, _) => logInfo(s"$e event. Still watching")
+              }
+          }
+        }
+      }
+      wo
+    }
+  }
+
+  private def cleanUpListener(source: BufferedSource, buffer: Buffer): Unit = {
+    buffer.close()
+    source.close()
+    break()
+  }
+
+  // Serves as a way to interrupt to the watcher thread.
+  // This closes the source the watcher is reading from and as a result triggers promise completion
+  def stopWatcher(): Unit = {
+    if (watchSource != null) {
+      buffer.close()
+      watchSource.close()
+    }
+  }
+
+  private def completeRequest(partialReq: Request.Builder): Request = {
+    kubeToken match {
+      case Some(tok) => partialReq.addHeader("Authorization", s"Bearer $tok").build()
+      case None => partialReq.build()
+    }
+  }
+
+  private def generateHttpUrl(urlSegments: Seq[String],
+    querySegments: Seq[(String, String)] = Seq.empty[(String, String)]): HttpUrl = {
+
+    val urlBuilder = HttpUrl.parse(kubeMaster).newBuilder
+    urlSegments map { pathSegment =>
+      urlBuilder.addPathSegment(pathSegment)
+    }
+    querySegments map {
+      case (query, value) => urlBuilder.addQueryParameter(query, value)
+    }
+    urlBuilder.build()
+  }
+
+  private def completeRequestWithExceptionIfNotSuccessful(
+    requestType: String,
+    response: Response,
+    additionalInfo: Option[Seq[String]] = None): Unit = {
+
+    if (!response.isSuccessful) {
+      response.body().close()
```

Comment:
It seems off to close the response body here since this private method is not what opened it. I think we should always close the response body in the parent method. We can still throw the exception from here, and as long as the calling methods close the response body in every circumstance (success or on exception) then we'll still be safe.